### PR TITLE
Feature/spark upgrade

### PR DESCRIPTION
**Pull request by rohitsinha54**
- CDAP uses the latest version of Spark  (1.2.0)
- Removed sparkContextStopBugFixer as this the bug in context close is fixed in this version.
- Tested stop functionality to work. 
Build : https://builds.cask.co/browse/CDAP-DUT426-1

### CODE REVIEW

**Comment by rohitsinha54** [comment]
Metrics test failed on build. I am running the build again.
Looks like something changed in spark metrics. Will do the needful fix on our side and update this PR.

**Comment by rohitsinha54** [comment]
Fixed the Metrics tests. New build running here: https://builds.cask.co/browse/CDAP-DUT426-3

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkProgramWrapper.java [hunk 1]
```diff
@@ -172,77 +167,19 @@ private void runUserProgram(SparkProgram sparkProgram) {
...
-          selectorClosed = true;
-          break;
-        } catch (IllegalAccessException iae) {
-          LOG.warn("Unable to access the selector field", iae);
-          throw Throwables.propagate(iae);
-        } catch (IOException ioe) {
-          LOG.info("Close on Selector threw IOException", ioe);
-          throw Throwables.propagate(ioe);
-        }
+    if (getSparkContext() != null) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkProgramWrapper.java]: So we no longer needs the hack?

File: cdap-app-fabric/src/main/java/co/cask/cdap/test/RuntimeStats.java [hunk 2]
```diff
@@ -161,4 +164,21 @@ public boolean apply(Map.Entry<String, AtomicLong> input) {
 
   private RuntimeStats() {
   }
+
+  /**
+   * Return the metrics key which which has the given starting and ending. If no such key is found returns an empty
+   * string.
+   *
+   * @param starting the starting part of the key
+   * @param ending   the ending part of the key
```
Comment (awholegunch) [inline cdap-app-fabric/src/main/java/co/cask/cdap/test/RuntimeStats.java]: ending the ending part of the key \[remove spaces; we don't align javadocs]

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkProgramWrapper.java (hunk 1 above)
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkProgramWrapper.java]: Nope. They removed the ConnectionManager class entirely.
I tested the Stop functionality to work without the hack and it works well.

**Comment by rohitsinha54** [comment]
Build :green_heart:

File: cdap-app-fabric/src/main/java/co/cask/cdap/test/RuntimeStats.java [hunk 3]
```diff
@@ -161,4 +164,21 @@ public boolean apply(Map.Entry<String, AtomicLong> input) {
...
   private RuntimeStats() {
   }
+
+  /**
+   * Return the metrics key which which has the given starting and ending. If no such key is found returns an empty
+   * string.
+   *
+   * @param starting the starting part of the key
+   * @param ending the ending part of the key
+   * @return the complete key if found else an empyty string
```
Comment (awholegunch) [inline cdap-app-fabric/src/main/java/co/cask/cdap/test/RuntimeStats.java]: an empty string

File: cdap-app-fabric/src/main/java/co/cask/cdap/test/RuntimeStats.java [hunk 4]
```diff
@@ -161,4 +164,21 @@ public boolean apply(Map.Entry<String, AtomicLong> input) {
 
   private RuntimeStats() {
   }
+
+  /**
+   * Return the metrics key which which has the given starting and ending. If no such key is found returns an empty
```
Comment (awholegunch) [inline cdap-app-fabric/src/main/java/co/cask/cdap/test/RuntimeStats.java]: Returns the metrics key having the given starting and ending parts. If no such key is found, returns an empty string.

**Comment by awholegunch** [comment]
One comment otherwise LGTM.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkProgramWrapper.java [hunk 5]
```diff
@@ -172,77 +167,19 @@ private void runUserProgram(SparkProgram sparkProgram) {
...
-        .env().blockManager().connectionManager();
-    }
-    return connectionManager;
   }
 
   /**
    * @return {@link SparkContext}
    */
-  public static SparkContext getSparkContext() {
+  private static SparkContext getSparkContext() {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkProgramWrapper.java]: You can remove this method entirely. I am not a big fan of having private accessor to access private field, unless you have addition logic in the accessor method.

**Comment by chtyim** [comment]
Just one comment left, otherwise LGTM.

**Comment by chtyim** [comment]
LGTM

**Comment by rohitsinha54** [comment]
Final build running here: https://builds.cask.co/browse/CDAP-DUT426-5

**Pull request merged**